### ISSUE TITLE
implement stun status effect

### ISSUE THREAD

**Issue by mysterymagination** [mysterymagination/adventures-of-mooty-wort#16]

**Comment by mysterymagination** [comment]
aaaalmost there in 7490ecc9a9f907911db66fc609a9700a2cd08175 (watta pain; I hadda hijack the enemy AI abl determination pipeline on one end and the anim pipeline on the other (although I coulda just added a real stunned animation...)) but I still see lingering last stun bolt icon intermittently lasting beyond the 0 tick

**Comment by mysterymagination** [comment]
stun's also arguably a little OP AF with 100% chance of two disrupted enemy turns.  The real problem is that you can almost recoup the whole cost (30) with two hero attacks (10+10), so I guess just make it more expensive? Maybe an HP cost to mix things up?

**Comment by mysterymagination** [comment]
here's a console log set from a case of lingering stun bolt icon:

```
// tries to load up non-existent FX JSON for placeholder PassStunner abl
XHR status says 404
// should be call to updateCharacterStatusStacks() at the bottom of handleEnemyTurnComplete()
mootywortrpgmech.js:471 stun has 1 ticks remaining
mootywortrpgmech.js:486 status stack; preparing to add stackdiv for stun, who still has 1 ticks left
// loading stun image async; meanwhile, beginNewRound is happening...
combat.js:113 ticking down The Yawning God's Stun to 0
alchemy.js:22 status effect reversal unset
combat.js:206 Dead players: 0, dead enemies: 0
characters.js:394 reached The Yawning God runAI fn... have fn!
characters.js:413 looking at player character with id mole
characters.js:416 the mole snuffles off to undefined...
combat.js:62 {"primordial_mandate":[0,30],"manyfold_embrace":[30,70],"pestilence":[70,90],"dark_star":[90,100]}
characters.js:515 The Yawning God's chosen abl is Primordial Mandate
// stun effect is seen as ticked all the way down so the view is done with it and the viewcontroller can tell the model to punt it
mootywortrpgmech.js:471 stun has 0 ticks remaining
mootywortrpgmech.js:473 breaktime
// async bolt image load completes and we wind up with an orphaned status icon
mootywortrpgmech.js:497 status stack; adding stackdiv for stun, who still has 0 ticks left, with offset dimens 38x311 to targetcanvascontainer with offset dimens 384x311
```

I think what's going on here is possibly a case of the async image load cb not firing until after another call through to processStatusEffectStack() has gone through and ticked stun down to 0 and removed it, basically resulting in it getting removed and then added back when the image loads.  If that's a possibility tho, why didn't it occur with other status icons?

**Comment by mysterymagination** [comment]
aha!  It doesn't only happen with Stun!
```
XHR status says 200
mootywortrpgmech.js:395 character sprite loaded
mootywortrpgmech.js:486 status stack; preparing to add stackdiv for bloodlust, who still has 1 ticks left
combat.js:113 ticking down The Yawning God's Bloodlust to 0
combat.js:206 Dead players: 0, dead enemies: 0
characters.js:394 reached The Yawning God runAI fn... have fn!
characters.js:413 looking at player character with id mole
characters.js:416 the mole snuffles off to undefined...
combat.js:62 {"primordial_mandate":[0,20],"manyfold_embrace":[20,50],"pestilence":[50,80],"dark_star":[80,100]}
characters.js:515 The Yawning God's chosen abl is Pestilence
mootywortrpgmech.js:497 status stack; adding stackdiv for bloodlust, who still has 0 ticks left, with offset dimens 24x274 to targetcanvascontainer with offset dimens 241x274
```
in this case I actually saw the new Stun icon draw over the Bloodlust icon (since Stun's icon has transparent areas), which should confirm that the Bloodlust effect was spliced out of the character's status effect array (the horizontal offset of the status icon stack is determined by the current status's index in the array so the only way overdraw like this should occur is if one of the icons isn't representing an active status effect in the array) and therefore the icon was truly orphaned.

also... hwhoops
```
status stack; adding stackdiv for stun, who still has 0 ticks left, with offset dimens 24x274 to targetcanvascontainer with offset dimens 241x274
mootywortrpgmech.js:502 status stack; effect with id stun was not found in the status effect array
mootywortrpgmech.js:504 status stack; stack div left is -24px based on offsetWidth 24 * effect index -1
```

Maybe the best solution to this problem would be to halt effective program flow until status icon image loads, a bit like we did in the case of spell FX -> pain animation flow, such that we can't get into these weird race conditions of {1 tick left -> start loading icon image -> tick down to 0 -> remove icon stack and associated status effect -> icon image finishes loading -> add icon stack w/ one icon that will linger forever}.  I guess that would involve passing a callback fn as arg into updateCharacterStatusStacks and forwarding it on to processStatusEffectStack?  That could get really ugly since we wanted to be able to call updateCharacterStatusStacks willy-nilly without worrying about impact on program flow.  Only obvious alternative would be limiting our calls to updateCharacterStatusStacks to avoid the current trouble case of handleEnemyTurnComplete -> updateCharacterStatusStacks inits loading of icon image -> stepCombat(state = beginNewRound) -> processRoundTop -> tick down status effect to 0 and remove it from data model -> updateCharacterStatusStacks removes relevant status stack div -> icon image finishes loading and re-adds its status stack.

Maybe not ideal solution, but a band-aid fix could be to just re-check effect ticks inside the image load cb to make sure we haven't hit this situation...

**Comment by mysterymagination** [comment]
well, the band-aid approach above seems ok and helps a lot with this problem, but I've still seen an orphaned stun icon sitting around occasionally.  Troublingly, I saw this log trace:
```
XHR status says 404
mootywortrpgmech.js:470 stun has 1 ticks remaining
mootywortrpgmech.js:480 status stack; removing stack div [object HTMLDivElement] with id yawning_god_stun_icon_stack
mootywortrpgmech.js:486 status stack; preparing to add stackdiv for stun, who still has 1 ticks left
combat.js:109 ticking down The Yawning God's Stun to 0
alchemy.js:22 status effect reversal unset
characters.js:508 The Yawning God's chosen abl is Manyfold Embrace, targeting Mooty Wort
mootywortrpgmech.js:470 stun has 0 ticks remaining
mootywortrpgmech.js:472 issue #16 breaktime
// why did we find the stack div here?  It was removed in the last pass at the end of the enemy turn and hasn't been re-added yet by the image load cb...
**mootywortrpgmech.js:480 status stack; removing stack div [object HTMLDivElement] with id yawning_god_stun_icon_stack**
mootywortrpgmech.js:523 status stack; status effect icon for stun is not needed as the effect has expired
```
the bolded line above shouldn't be since the stack div had already been removed and shouldn't have been re-added anyplace...

**Comment by mysterymagination** [comment]
also got a probably related cthulhu-crazy-and-not-in-the-good-way bug where sometimes the mole seems to either have the yawning god's abilities to choose from and/or the showSpellEffect() function for him will look for a YG abl, not find it, and then throw  Uncaught (in promise) TypeError: Cannot read property 'fxDataFileName' of undefined

**Comment by mysterymagination** [comment]
yeah, now we've got some bizarre multi-turn rapid fire madness happening with people taking each other's abilities and all sorts of crap.  My poor little state machine!

**Comment by mysterymagination** [comment]
possible state machine frotzers:

1. Early jump out of anim pipeline if/when FX data file load fails (expected for the placeholder abl PassStunned)

**Comment by mysterymagination** [comment]
muuuuuh... on Windows Chrome 86 I'm not seeing the issue either on localhost or from GitHub Pages deployment.  Wtf?  The state is nuts on Android and I'm pretty sure I saw some of the same trouble on a Linux run on localhost and/or Pages.  I did still see a lingering status icon thingy on Windows tests tho

Edit: I don't see the problem on Android when I check the desktop site button, and I'd bet that I was using mobile mode while testing on Linux.  Dunno what mobile web might have to do with anything, but I'll take repro!

**Comment by mysterymagination** [comment]
looks like somehow we're never setting combat state to beginNewRound after the first iteration of runEnemy; instead we runEnemy twice and then go to beginNewRound, but only in mobile UI.

**Comment by mysterymagination** [comment]
ok, so the problem with doubled up enemy turns seems to be a weird one --
1. in handleEnemyTurnComplete, we search for a next living character in enemy array starting after the index in that array of the current turn owner (expected to be the active enemy).  For some reason tho, the currentTurnOwner is the mole.
2. The mole is not found in enemy array at all, returning index -1.
3. findFirstLivingCharacter checks to make sure array length is >= afterIndex (-1 from above) + 2.  It is length 1, so it passes.  We then look for a character in afterIndex+1 and The Yawning God is found at index 0!  Huzzah, it is the Yawning God's turn... again.
4. Next iteration of handleEnemyTurnComplete has currentTurnOwner set to Yawning God and things work as expected.

So the problem is that we're searching for the mole in the enemy party because the currentTurnOwner was not handed over to Yawning God at the beginning of his (first) turn (or something)

Seems that handlePlayerTurnComplete is handing off turn owership to YG correctly, and in first handleEnemyTurnComplete it is back at mole so the fault is somewhere in between?

oooh ok so basically what's happening is first the reverse of the above wrt turns, with handlePlayerTurnComplete coming up when the turn owner is actually YG.  Same -1 index thing leading to erroneous reass of currentTurnOwner to the next/only living [player] happens.  This hiccup occurs because we somehow managed to call handlePlayerTurnComplete first when the actual state machine proc finished with player and then again, later, after the turn ownership had been handed over to enemy by the first iteration of handlePlayerTurnComplete, when showSpellEffect returns after several rounds of async callbacks.

We definitely enter handlePlayerTurnComplete twice when I open with a single Burrow Furrowed Brow casting; the first time assigns turn ownership to YG as expected but then the second bugs out and assigns it back to mole, as explained above.  Both call stacks only have playPainAnim as the caller of handlePlayerTurnComplete, so my money's on some async shenanigans in showSpellEffect and/or playPainAnim that allows the state-forwarding callbackFunction they're given to be called twice.
->We're calling playPainAnimation twice somehow

**Comment by mysterymagination** [comment]
aha ahhhaaha!  It's the bloody touch/mouse events!  I forgot all about that crap, but it seems I'm getting both mouseup and touchend callbacks (which both call my handler function touchUp and start the showSpellEffect flow) for a single command button press in mobile mode.  I thought the whole reason we needed both touch and mouse was because only one would fire depending on whether you were in desktop or mobile web... whelp

hmm ok so apparently the reason I added both touch and mouse wasn't to handle clicking as mouse handles that fine on both desktop and mobile -- it was to allow for long-click detection and the abl description printing.  On desktop, we get our mousedown when long-clicking but not on mobile.  On mobile we need touchstart and touchend to detect long-click.  So I guess the solution is to make a dedicated touch handler that only handles a detected long-click and DOES NOT kick off the showSpellEffect flow; that can safely be left to mouseup/down?

**Comment by mysterymagination** [comment]
maybe fixed in 43c52519fee98309e3c8997fc8f7f780bb93e501 ?  Haven't seen the lingering icon again yet, and that nasty input handling subbug here was royally messing up the state machine...
